### ISSUE TITLE
Inference outside Unity

### ISSUE THREAD

**Issue by Vanoo** [Unity-Technologies/ml-agents#1959]
Hi, i'm currently using your's wonderful ml-agents as a training environment for a robot to pick some object and move them at a specific destination, in the hope that I can use this neural network in a real robot.

I use a discrete action space, with 2 branches of size 2.

The problem that I encounter is that I don't understand which value to put for the inference, and I need to make sense of the output that I get.

Here my code :
```
pendulumA_rotation = [1,5,1,1]
pendulumB_rotation = [1,1,1,1]
current_goal_position = [1,1,1]
gripPoint_position = [1,1,1]
phase = [1]

vectorObservation = pendulumA_rotation + pendulumB_rotation + current_goal_position + gripPoint_position + phase
actionMask = [1,1,1,1]

graph = tf.train.import_meta_graph(metaName)
with tf.Session() as sess:
      graph.restore(sess, tf.train.latest_checkpoint('./'))
      y_out = sess.run(['action:0'], feed_dict={'vector_observation:0': [vectorObservation],
                                                                  'action_masks:0': [actionMask]})
      print(y_out)
```
with those values i get in action:0 : [[-0.52742714, -0.8918941 , -0.53989583, -0.874211 ]]

Question :
Did action mask really contain boolean value, true(1) for movement possible and false(0) for impossible movement?
Did Each element of the output corresponds to a probability for a specific action? And if so for each branch unity take the indices and put it in vectorAction[], with the example here, I get vectorAction = [0,0], is that correct ?

Thx for your help and the time that you can put on my problem.

Sorry for my english it's not my first language.

**Comment by Vanoo** [comment]
> Hi, i'm currently using your's wonderful ml-agents as a training environment for a robot to pick some object and move them at a specific destination, in the hope that I can use this neural network in a real robot.
> 
> I use a discrete action space, with 2 branches of size 2.
> 
> The problem that I encounter is that I don't understand which value to put for the inference, and I need to make sense of the output that I get.
> 
> Here my code :
> 
> ```
> pendulumA_rotation = [1,5,1,1]`
> pendulumB_rotation = [1,1,1,1]
> current_goal_position = [1,1,1]
> gripPoint_position = [1,1,1]
> phase = [1]
> 
> vectorObservation = pendulumA_rotation + pendulumB_rotation + current_goal_position + gripPoint_position + phase
> actionMask = [1,1,1,1]
> 
> graph = tf.train.import_meta_graph(metaName)
> with tf.Session() as sess:
> graph.restore(sess, tf.train.latest_checkpoint('./'))
> y_out = sess.run(['action:0'], feed_dict={'vector_observation:0': [vectorObservation],
> 'action_masks:0': [actionMask]})
> print(y_out)
> ```
> 
> with those values i get in action:0 : [[-0.52742714, -0.8918941 , -0.53989583, -0.874211 ]]
> 
> Question :
> Did action mask really contain boolean value, true(1) for movement possible and false(0) for impossible movement?
> Did Each element of the output corresponds to a probability for a specific action? And if so for each branch unity take the indices and put it in vectorAction[], with the example here, I get vectorAction = [0,0], is that correct ?
> 
> Thx for your help and the time that you can put on my problem.
> 
> Sorry for my english it's not my first language.

**Comment by vincentpierre** [comment]
As of version v0.8, the [action node](https://github.com/Unity-Technologies/ml-agents/blob/0.8.1/ml-agents/mlagents/trainers/models.py#L345) is normalized **log** probabilities. To get the action, you will need to use a multinomial operator to sample an action with each of the two branches.

**Comment by Vanoo** [comment]
Thx you for the fast answer. 
I will look on multinomial operator as i don't know what that is.

For my first question about the action_mask, it is correct ?

**Comment by vincentpierre** [comment]
Yes, I means the action is valid. You can test this to be sure.

**Comment by vincentpierre** [comment]
Hi, I am not sure I understand that last comment,
Multinomial is a sampling operator. Softmax will give you probabilities for each actions and multinomial is what will sample the action according to the probability. It is true that there will still be some randomness but always according to the probabilities "decided" by the agent. This randomness is crucial for exploration.
> if so it can explain some strange result that i saw when the reward stagnate to a certain point during multiple step just to drop significantly after.
There can be multiple causes for such a drop, like a change in the environment during training.

**Comment by Vanoo** [comment]
Hi, i was saying that with just the result of the softmax, the best result is the one with the highest probability, and without change in the network it's always the same result.
With the multinomial operator we get a result base on a dice roll with the output of our network as the probabilities of the dice.

In the end i did this to convert the output to a onehot encoding of the action.
```
        y_out = sess.run(['action:0'], feed_dict={'vector_observation:0': [vectorObservation],
                                                                    'action_masks:0': [actionMask]})
        outputs = y_out[0][0]

        prob_action1 = outputs[:dimAction1]
        prob_action2 = outputs[dimAction2:]

        norm_action1 = [float(i)/sum(prob_action1) for i in prob_action1]
        norm_action2 = [float(i)/sum(prob_action2) for i in prob_action2]

        action1 = np.random.multinomial(1,norm_action1)
        action2 = np.random.multinomial(1,norm_action2)
```
